File: Android/src/app/src/main/java/com/google/ai/edge/gallery/customtasks/exampleagent/ExampleAgentTools.kt
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.ai.edge.gallery.customtasks.exampleagent

import android.util.Log
import com.google.ai.edge.litertlm.Tool
import com.google.ai.edge.litertlm.ToolParam
import com.google.ai.edge.litertlm.ToolSet
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

private const val TAG = "AGExampleAgentTools"

/**
 * TEMPLATE: a set of local function-calling tools.
 *
 * This is the simplest way to give an on-device LLM new abilities: expose plain Kotlin functions
 * annotated with [@Tool][Tool]. When the model decides a function is needed, the LiteRT-LM runtime
 * calls it for you and feeds the returned value back into the conversation.
 *
 * How it works:
 * - The class implements [ToolSet].
 * - Each callable function is annotated with `@Tool(description = ...)`. The description is what the
 *   model reads to decide *when* to call it — keep it clear and specific.
 * - Each parameter is annotated with `@ToolParam(description = ...)`. Use simple types
 *   (String, Int, Double, Boolean).
 * - The function returns a `Map<String, ...>` (or any serializable value). The map is sent back to
 *   the model as the tool result; include a `result`/`error` + `status` so the model can reason
 *   about success/failure.
 * - Wrap an instance with `com.google.ai.edge.litertlm.tool(...)` and pass it to
 *   `LlmChatModelHelper.initialize(tools = listOf(tool(ExampleAgentTools(...))), ...)`.
 *
 * Guidelines:
 * - Keep tools fast and side-effect-light, or do heavy work off the calling thread. Tool functions
 *   are invoked by the runtime synchronously during generation.
 * - Validate inputs and return a structured error instead of throwing.
 * - Use the [onToolCalled] callback to surface activity to your UI (progress, undo, analytics).
 *
 * To add your own tool: copy one of the functions below, rename it, write a precise description, and
 * implement the body. That's it — the model can now call it.
 *
 * @param onToolCalled Optional callback invoked whenever a tool runs, with a short human-readable
 *   label. Use it to drive a "tool activity" indicator in your UI.
 */
class ExampleAgentTools(private val onToolCalled: (label: String) -> Unit = {}) : ToolSet {

  /** A no-argument tool. */
  @Tool(description = "Returns the current local date and time.")
  fun getCurrentDateTime(): Map<String, String> {
    onToolCalled("Reading the current time")
    @Suppress("JavaTimeDefaultTimeZone") val now = LocalDateTime.now()
    val formatted = now.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))
    Log.d(TAG, "getCurrentDateTime -> $formatted")
    return mapOf("result" to formatted, "status" to "succeeded")
  }

  /** A tool with typed parameters that returns a computed value (and a structured error). */
  @Tool(
    description =
      "Performs a basic arithmetic operation on two numbers and returns the result. " +
        "Supported operations: add, subtract, multiply, divide."
  )
  fun calculate(
    @ToolParam(description = "The operation: one of add, subtract, multiply, divide.")
    operation: String,
    @ToolParam(description = "The first operand.") a: Double,
    @ToolParam(description = "The second operand.") b: Double,
  ): Map<String, Any> {
    onToolCalled("Calculating $a $operation $b")
    val result =
      when (operation.trim().lowercase()) {
        "add",
        "+" -> a + b
        "subtract",
        "-" -> a - b
        "multiply",
        "*" -> a * b
        "divide",
        "/" -> {
          if (b == 0.0) {
            return mapOf("error" to "Cannot divide by zero", "status" to "failed")
          }
          a / b
        }
        else ->
          return mapOf(
            "error" to "Unknown operation: \"$operation\"",
            "status" to "failed",
          )
      }
    Log.d(TAG, "calculate($operation, $a, $b) -> $result")
    return mapOf("result" to result, "status" to "succeeded")
  }

  /**
   * A tool that does real work but stays self-contained for the template. Replace the body with a
   * call into your repository / Android API / network client as needed.
   */
  @Tool(description = "Counts the number of words in the given text.")
  fun countWords(
    @ToolParam(description = "The text whose words should be counted.") text: String
  ): Map<String, Any> {
    onToolCalled("Counting words")
    val count = text.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }.size
    return mapOf("result" to count, "status" to "succeeded")
  }
}

File: iOS/Gallery/CustomTasks/ExampleAgent/ExampleAgentTools.swift
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Port of customtasks/exampleagent/ExampleAgentTools.kt
//
// TEMPLATE: a set of local function-calling tools.
//
// NOTE: Android used @Tool / @ToolParam / ToolSet annotations from litertlm for
// reflection-based registration. iOS represents tools as JSON function-spec
// dictionaries. The class below mirrors the same three tool functions and
// exposes the JSON specs consumed by ExampleAgentTask for LlmModelHelper.

import Foundation

/// TEMPLATE: local tool set for the ExampleAgent.
/// Mirrors `ExampleAgentTools`.
///
/// Guidelines (same as Android):
/// - Keep tools fast and side-effect-light.
/// - Validate inputs; return a structured error instead of throwing.
/// - Use `onToolCalled` to surface activity to the UI.
final class ExampleAgentTools {
    private let onToolCalled: (String) -> Void

    init(onToolCalled: @escaping (String) -> Void = { _ in }) {
        self.onToolCalled = onToolCalled
    }

    // MARK: - Tool functions

    /// Returns the current local date and time.
    func getCurrentDateTime() -> [String: String] {
        onToolCalled("Reading the current time")
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        let formatted = formatter.string(from: Date())
        return ["result": formatted, "status": "succeeded"]
    }

    /// Performs a basic arithmetic operation on two numbers.
    func calculate(operation: String, a: Double, b: Double) -> [String: Any] {
        onToolCalled("Calculating \(a) \(operation) \(b)")
        let result: Double
        switch operation.trimmingCharacters(in: .whitespaces).lowercased() {
        case "add", "+":      result = a + b
        case "subtract", "-": result = a - b
        case "multiply", "*": result = a * b
        case "divide", "/":
            guard b != 0 else {
                return ["error": "Cannot divide by zero", "status": "failed"]
            }
            result = a / b
        default:
            return ["error": "Unknown operation: \"\(operation)\"", "status": "failed"]
        }
        return ["result": result, "status": "succeeded"]
    }

    /// Counts the number of words in the given text.
    func countWords(text: String) -> [String: Any] {
        onToolCalled("Counting words")
        let count = text.trimmingCharacters(in: .whitespacesAndNewlines)
            .components(separatedBy: .whitespacesAndNewlines)
            .filter { !$0.isEmpty }
            .count
        return ["result": count, "status": "succeeded"]
    }

    // MARK: - JSON function specs
    //
    // NOTE: These mirror @Tool + @ToolParam annotations. The LlmModelHelper stub
    // ignores them; a real MediaPipe/LiteRT bridge would use them to build the
    // function-calling context for the model.

    static var functionSpecs: [[String: Any]] {
        [
            [
                "name": "getCurrentDateTime",
                "description": "Returns the current local date and time.",
                "parameters": ["type": "object", "properties": [:], "required": []],
            ],
            [
                "name": "calculate",
                "description": "Performs a basic arithmetic operation on two numbers and returns the result. Supported operations: add, subtract, multiply, divide.",
                "parameters": [
                    "type": "object",
                    "properties": [
                        "operation": ["type": "string",
                                      "description": "The operation: one of add, subtract, multiply, divide."],
                        "a": ["type": "number", "description": "The first operand."],
                        "b": ["type": "number", "description": "The second operand."],
                    ],
                    "required": ["operation", "a", "b"],
                ],
            ],
            [
                "name": "countWords",
                "description": "Counts the number of words in the given text.",
                "parameters": [
                    "type": "object",
                    "properties": [
                        "text": ["type": "string",
                                 "description": "The text whose words should be counted."],
                    ],
                    "required": ["text"],
                ],
            ],
        ]
    }

    // MARK: - Dispatch

    func dispatch(functionName: String, args: [String: Any]) -> String {
        var result: [String: Any]
        switch functionName {
        case "getCurrentDateTime":
            result = getCurrentDateTime()
        case "calculate":
            result = calculate(
                operation: args["operation"] as? String ?? "",
                a: args["a"] as? Double ?? 0,
                b: args["b"] as? Double ?? 0
            )
        case "countWords":
            result = countWords(text: args["text"] as? String ?? "")
        default:
            result = ["error": "Unknown function", "status": "failed"]
        }
        let data = (try? JSONSerialization.data(withJSONObject: result)) ?? Data()
        return String(data: data, encoding: .utf8) ?? "{}"
    }
}
